torch/csrc/autograd/function.h

```diff
@@ -39,79 +43,50 @@ struct edge_hasher {
-struct FunctionFlags {
-  // Roughly speaking, is_executable corresponds to requires_grad.
-  // It's true if any input requires grad and gradient calculation is enabled.
-  // See http://pytorch.org/docs/notes/autograd.html for more details.
-  bool is_executable = false;
-  // What functions take the output of this function as input.
-  // There is one function per output of this function.
-  function_list next_functions;
-};
-
 namespace detail {
-
-// Why can't we just combine the set_variable and set_tensor variants
-// into one set of overloads?  The problem is Variable is convertible
-// to both Tensor and ArrayRef<Variable>, making the overload ambiguous.
-
-// Invariant: this function unconditionally calls f.next_functions.emplace_back
-inline void set_function_flags(FunctionFlags& f, const Variable& var) {
-  if (!var.defined()) {
-    f.next_functions.emplace_back();
-    return;
-  }
-  f.is_executable |= var.requires_grad();
-  if (var.grad_fn()) {
-    f.next_functions.emplace_back(var.grad_fn(), var.output_nr());
-  } else if (var.requires_grad()) {
-    f.next_functions.emplace_back(var.grad_accumulator(), 0);
-  } else {
-    f.next_functions.emplace_back();
+inline edge_type make_edge(const Variable &variable) {
+  if (variable.defined()) {
+    if (variable.grad_fn() != nullptr) {
+      return {variable.grad_fn(), variable.output_nr()};
+    } else if (variable.requires_grad()) {
+      return {variable.grad_accumulator(), 0};
+    }
   }
+  return {nullptr, 0};
 }
 
-struct SetFunctionFlags : IterArgs<SetFunctionFlags> {
-  FunctionFlags& out;
-  SetFunctionFlags(FunctionFlags& out) : out(out) {}
-  using IterArgs<SetFunctionFlags>::operator();
-  void operator()(const Variable& v) { set_function_flags(out, v); }
-};
-
-struct SetTensorFunctionFlags : IterArgs<SetTensorFunctionFlags> {
-  FunctionFlags& out;
-  SetTensorFunctionFlags(FunctionFlags& out) : out(out) {}
-  using IterArgs<SetTensorFunctionFlags>::operator();
-  void operator()(const Tensor& t) {
-    set_function_flags(out, static_cast<const Variable&>(t));
+struct MakeNextFunctionList : IterArgs<MakeNextFunctionList> {
+  function_list next_functions;
+  using IterArgs<MakeNextFunctionList>::operator();
+  void operator()(const Variable& variable) {
+    next_functions.push_back(make_edge(variable));
   }
 };
+} // namespace detail
 
+// Returns true if any of the variables in the list require a gradient.
+inline bool any_variable_requires_grad(const variable_list& variables) {
+  return std::any_of(
+      variables.begin(), variables.end(), [](const Variable& variable) {
+        return variable.requires_grad();
+      });
+}
 
-} // namespace detail
+template <typename... Variables>
+function_list get_next_functions(Variables&&... variables) {
+  if (!GradMode::is_enabled()) return {};
+  detail::MakeNextFunctionList make;
+  make.apply(std::forward<Variables>(variables)...);
+  return std::move(make.next_functions);
```

Comment:
RVO will not take place for members. RVO (Named RVO to be precise) will only take place in very simple cases, basically when the value returned is declared on its own and has the same type as the return type (`X x; ...; return x;`). From [cppreference](http://en.cppreference.com/w/cpp/language/copy_elision):
> If a function returns a class type by value, and the return statement's expression is the name of a non-volatile object with automatic storage duration ... then copy/move (since C++11) is omitted.

I.e. the return expression literally has to be the name of a value of the type. Otherwise a copy will occur.
https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization/12953129